standard/expressions.md

```diff
@@ -3875,6 +3876,74 @@ A lifted ([§12.4.8](expressions.md#1248-lifted-operators)) form of the range op
 
 The range operator is non-associative ([§12.4.2](expressions.md#1242-operator-precedence-and-associativity)).
 
+## §switch-expression-new-clause Switch expression
+
+A *switch_expression* provides `switch`-like semantics in an expression context.
+
+```ANTLR
+switch_expression
+    : range_expression
+    | switch_expression 'switch' '{' (switch_expression_arms ','?)? '}'
+    ;
+switch_expression_arms
+    : switch_expression_arm
+    | switch_expression_arms ',' switch_expression_arm
+    ;
+switch_expression_arm
+    : pattern case_guard? '=>' switch_expression_arm_expression
+    ;
+switch_expression_arm_expression
+    : expression
+    ;
+```
+
+A *switch_expression* is not an *expression_statement*.
+
+There is a *switch expression conversion* (§switch-expression-conversion) from a switch expression to a type `T`
+if there is an implicit conversion from every *switch_expression_arm_expression* of each of the switch expression's *switch_expression_arm*s to `T`.
+
+If a switch expression is not subject to a *switch expression conversion*, then
+
+- The type of the *switch_expression* is the best common type [§12.6.3.15](expressions.md#126315-finding-the-best-common-type-of-a-set-of-expressions)) of the *switch_expression_arm_expression*s of the *switch_expression_arm*s if such a type exists and each *switch_expression_arm_expression* can be implicitly converted to that type.
+- It is an error if no such type exists.
+
+It is an error if some *switch_expression_arm*'s pattern cannot affect the result because some previous pattern and guard will always match.
```

Comment:
Is it not that some previous arm will always match but that the sum of all the previous arms match the whole domain, i.e. an arm is *subsumed* (§11.3) as with the switch statement (§13.8.3)?

Or is this not subsumption as it includes the guards?

Or does pattern subsumption (§11.3) need to be expanded to include guards?

Even if it is not subsumption I think it needs to be clear that it is the sum of *all* preceding arms.

Reply:
This should probably be restated in terms of subsumption. case expression arms with guards are only considered if the guard is a constant expression equal to true (if it's equal to false, it doesn't subsume).